File: examples/Builder.py
#!/usr/bin/env python

# ----------
# Builder.py
# ----------

# http://en.wikipedia.org/wiki/Builder_pattern

class Room (object) :
    pass

class EnchantedRoom (Room) :
    pass

class Door (object) :
    def __init__ (self, r, s) :
        self.__r = r
        self.__s = s

class EnchantedDoor (Door) :
    def __init__ (self, r, s) :
        Door.__init__(self, r, s)

class Maze (object) :
    def __init__ (self) :
        self.__rooms = []
        self.__doors = []

    def add_room (self, r) :
        self.__rooms += [r]

    def add_door (self, d) :
        self.__doors += [d]

    def room (self, i) :
        return self.__rooms[i]

    def door (self, i) :
        return self.__doors[i]

class MazeBuilder (object) :
    def __init__ (self) :
        self.__maze = Maze()

    def build_room (self) :
        self.__maze.add_room(Room())

    def build_door (self, r, s) :
        self.__maze.add_door(Door(r, s))

    def maze (self) :
        return self.__maze

class EnchantedMazeBuilder (MazeBuilder) :                # object?
    def __init__ (self) :
        MazeBuilder.__init__(self)

    def build_room (self) :
        self._MazeBuilder__maze.add_room(EnchantedRoom())

    def build_door (self, r, s) :
        self._MazeBuilder__maze.add_door(EnchantedDoor(r, s))

def create_maze (mb) :
    mb.build_room()
    mb.build_room()
    mb.build_room()
    mb.build_door(mb.maze().room(0), mb.maze().room(1))
    mb.build_door(mb.maze().room(1), mb.maze().room(2))
    return mb.maze()

print "Builder.py"

mb = MazeBuilder()
m  = create_maze(mb)
assert type(mb)        is MazeBuilder
assert type(m)         is Maze
assert type(m.room(0)) is Room
assert type(m.door(0)) is Door

mb = EnchantedMazeBuilder()
m  = create_maze(mb)
assert type(mb)        is EnchantedMazeBuilder
assert type(m)         is Maze
assert type(m.room(0)) is EnchantedRoom
assert type(m.door(0)) is EnchantedDoor

print "Done."

File: examples/Builder.java
// ------------
// Builder.java
// ------------

// http://en.wikipedia.org/wiki/Builder_pattern

import java.util.ArrayList;
import java.util.List;

class Room {}
class EnchantedRoom extends Room {}

class Door {
    private Room _r;
    private Room _s;

    public Door (Room r, Room s) {
        _r = r;
        _s = s;}}

class EnchantedDoor extends Door {
    public EnchantedDoor (Room r, Room s) {
        super(r, s);}}

final class Maze {
    private List<Room> _rooms = new ArrayList<Room>();
    private List<Door> _doors = new ArrayList<Door>();

    public void addRoom (Room r) {
        _rooms.add(r);}

    public void addDoor (Door d) {
        _doors.add(d);}

    public Room room (int i) {
        return _rooms.get(i);}

    public Door door (int i) {
        return _doors.get(i);}}

class MazeBuilder {
    protected Maze _maze;

    public MazeBuilder () {
        _maze = new Maze();}

    public void buildRoom () {
        _maze.addRoom(new Room());}

    public void buildDoor (Room r, Room s) {
        _maze.addDoor(new Door(r, s));}

    public Maze maze () {
        return _maze;}}

class EnchantedMazeBuilder extends MazeBuilder {
    public void buildRoom () {
        _maze.addRoom(new EnchantedRoom());}

    public void buildDoor (Room r, Room s) {
        _maze.addDoor(new EnchantedDoor(r, s));}}

abstract class Game {
    public static Maze createMaze (MazeBuilder mb) {
        mb.buildRoom();
        mb.buildRoom();
        mb.buildRoom();
        mb.buildDoor(mb.maze().room(0), mb.maze().room(1));
        mb.buildDoor(mb.maze().room(1), mb.maze().room(2));
        return mb.maze();}}

final class Builder {
    public static void main (String[] args) {
        System.out.println("Builder.java");

        {
        MazeBuilder mb = new MazeBuilder();
        Maze        m  = Game.createMaze(mb);
        assert mb.getClass()        == MazeBuilder.class;
        assert m.getClass()         == Maze.class;
        assert m.room(0).getClass() == Room.class;
        assert m.door(0).getClass() == Door.class;
        }

        {
        MazeBuilder mb = new EnchantedMazeBuilder();
        Maze        m  = Game.createMaze(mb);
        assert mb.getClass()        == EnchantedMazeBuilder.class;
        assert m.getClass()         == Maze.class;
        assert m.room(0).getClass() == EnchantedRoom.class;
        assert m.door(0).getClass() == EnchantedDoor.class;
        }

        System.out.println("Done.");}}
